mslaved/mslaved.8
.TH MSLAVED 8 "" "MandelSpawn"
.SH NAME 
mslaved \- MandelSpawn computation server
.SH SYNOPSIS  
.B mslaved 
[
.B \-t
timeout ] [
.B \-n
niceval ]
.SH DESCRIPTION  
.I mslaved
is the computation server daemon for MandelSpawn (see 
.I xms (1)
and
.IR bms (1)).
It accepts requests for calculating rectangles of pixels on its 
standard input, which should be a datagram socket.  
Normally \fImslaved\fP is started on demand
by 
.IR inetd (8C);
it may also be started manually using 
.IR mslavedc (8),
as in
.PP
	mslavedc ./mslaved
.PP
To spawn 
.I mslaved
processes on multiple machines, use
.IR enslave (1).
.PP
Note that
.I mslaved 
can \fInot\fP be started directly from the shell without using
.I mslavedc
or
.IR enslave .
.PP
The \fB\-n\fP option specifies a nice value; if no \fB\-n\fP option
is present, nice 10 will be used.  Use \fB\-n0\fP for no nice.
The \fB\-t\fP option specifies a timeout: \fImslaved\fP will exit
if no calculation requests are
received within the given number of seconds.  The default
timeout is 60 seconds, \fB\-t0\fP means never time out.
.SH "SEE ALSO"
xms(1), bms(1), enslave(1), mslavedc(8)
.PP
.SH COPYRIGHT
.if n Copyright (C) 1990, 1991 Andreas Gustafsson (gson@niksula.hut.fi)
.if t Copyright \(co 1990, 1991 Andreas Gustafsson (gson@niksula.hut.fi)


ChangeLog
Wed Aug  6 17:23:33 1997  Andreas Gustafsson  <gson@guava.araneus.fi>

	* Wrapped version 0.07pl2

	* Switched from compress to gzip.

	* Updated contact information in README.

	* Commented out the X11R[23] support in xms/Imakefile
	because it misidentified X11R6 as X11R[23], and no-one
	uses X11R[23] anymore anyway.

	* 0.07pl1 fixed select() bug that slowed down bms to one packet
        per second.

Sat Sep  4 03:08:49 1993  Andreas Gustafsson  (gson@guano)

	* Wrapped version 0.07

	* Fixed two bugs in Ms.c which caused core dumps under
	Linux

Sun Aug 15 12:11:38 1993  Andreas Gustafsson  (gson@guano)

	* Added Linux FNDELAY hack
	
	* Added strdup() replacement (needed by pure 4.3BSD)

Sun Jun  6 04:27:00 1993  Andreas Gustafsson  (gson@guano)

	* Rewrote parts of the README file.

	* Gave up futile attempts to declare malloc() using system 
	include files.

	* Combined the various INSTALL files into a single file.

	* Fixed an use of freed memory and several memory leaks.

Sat May 22 03:04:12 1993  Andreas Gustafsson  (gson@guano)

	* Updated years in copyright messages.

Fri May  7 04:41:48 1993  Andreas Gustafsson  (gson@guano)

	* Version 0.06g ported to HP700-series and DEC Alpha
	machines.  The rubberband box can now be manipulated
	while the calculation is in progress.

	* Version 0.06h was put under RCS.  This necessitated a 
	restructuring of the code into a more traditional 
	directory tree with no links between directories.  There
	is now a lib/ directory for code shared between the
	different programs.  PPM support was added to bms.

	* Version 0.06f introduced experimental Henon fractal code
	in mslaved and associated generic parameter decoding code.

	* Version 0.06d1 changed to support a BitMapUnit of 32 in
	b/w operation.  Maybe it was a mistake to remove the
	8-bit code.

Thu May 14 19:58:16 1992  Andreas Gustafsson  (gson@psykoman.cs.hut.fi)

	* Lots of changes to support 24-bit visuals

Sat Feb  8 21:15:44 1992  Andreas Gustafsson  (gson at gunnar)

	* Added support for pipe operation in mslaved.c; introduced
	HAVE_SOCKETS and HAVE_SYSLOG compile-time flags

Wed Feb  5 16:57:41 1992  Andreas Gustafsson  (gson at spiff.cs.hut.fi)

	* Made the visual matching code	use the default visual for 
	black-and-white operation

	* Added colormap animation

Fri Dec  6 17:32:39 1991  Andreas Gustafsson  (gson at bulldozer.cs.hut.fi)

	* Allowed empty lines and # comments in .enslave

Sun Dec  1 16:18:26 1991  Andreas Gustafsson  (gson at punisher.cs.hut.fi)

	* New development version 0.06a.  Includes visual matching code in
	Mama.c, and new option handling in mslaved and mslavedc.

Sun Nov 24 12:38:12 1991  Andreas Gustafsson  (gson at hydroman.cs.hut.fi)

	* Wrapped version 0.06

	* Added option details to the synopsis in xms.man

	* Tested the new ms_draw() on monochrome and 8-bit colour
	Suns.

	* Completely rewrote the black-and-white special case display
	support in Ms.c:ms_draw(), so that it now constructs the image
	a byte at a time rather than a word at a time.  This should
	avoid any byteorder problems.

Tue Nov  5 17:23:38 1991  Andreas Gustafsson  (gson at bulldozer.cs.hut.fi)

	* Version number is now 0.05p.  Changed some key bindings to be 
	more consistent with the new "zoom by default" policy, and
	updated the manual page correspondingly.

Mon Sep 30 01:01:40 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added XtNinput argument to shell widget; that ought to
	fix the keyboard focus problem with OpenWindows (not tested).

Sun Sep 22 12:07:27 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Edited README and mslaved/Makefile; now gcc is stronly recommended
	only for old CISCs.  Made floating point the default for MIPS in
	ms_real.h as suggested by broadley's benchmarks.

	* restored the menu entries for using the old window but made the
	popup alternative more defaultish (no, I don't run spell on
	ChangeLogs). 

	* added #include <sys/file.h> to ms_ipc.h (needed for FNDELAY in
	HP-UX) 

	* fixed some syntax errors in mslaved.c

	* added #include <netinet/in.h> in Ms.c to get htonl() and ntohs()
	macro definitions (needed for HP-UX)

Sat Sep 14 00:09:20 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* #ifdef'ed away the menu entries for reusing the old window

Sun Aug 11 14:42:55 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added "inverse box zooming" as suggested by RMS.  Still
	untested.

Sat Aug 10 10:32:51 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Added #include <errno.h> in work.c

	* Wrapped 0.05j and mailed to broadley for beta testing

	* Renamed split.c to mspawn.c

	* Tested 0.05h on SPARCstation IPC.  Fixed some bugs in the
	noImakefile and documented problems with missing imake and
	Aw libs.  Version with these fixes called 0.05i.

Fri Jul 26 14:23:57 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Fixed some more typos and made some other changes in the man
	pages

	* Wrapped 0.05h	and gave it to ufa@niksula.hut.fi for beta 
	testing 

Sun Jul 21 12:36:22 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Wrote the bms batch mode client and renamed the X client to
	xms.  This also caused lots of changes to both source and 
	documentation of xms because of extensive code sharing between the
	two clients.  Mandelbrot-dependent but X-independent
	code has been separated into a new module, mspawn.c, leaving
	code that is both X- and Mandelbrot-independent in work.c.
	The timeout system has been completely rewritten.

	* MIPS fixed-point code has been tested on a DECStation.

Wed Jun 26 17:01:04 1991  Andreas Gustafsson  (gson at gunnar)

	* Separated the X-independent workload distribution stuff
	into a new module, work.c.  Renamed all X-independent files
	to lowercase names to reduce ugliness.  Added missing "2" in
	MIPS fracmult2 function name; need to retest MIPS code before
	distributing it.

Tue Jun 18 21:34:57 1991  Andreas Gustafsson  (gson at spica.hut.fi)

	* Fixed several typos in the man page


mslaved/mslaved.c
/* mslaved.c - MandelSpawn computation server deamon */

/*  
   This file is part of MandelSpawn, a network Mandelbrot program.

    Copyright (C) 1990-1993 Andreas Gustafsson

    MandelSpawn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License, version 1,
    as published by the Free Software Foundation.

    MandelSpawn is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License,
    version 1, along with this program; if not, write to the Free 
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <signal.h>
#ifdef HAVE_SYSLOG
#include <syslog.h>
#endif

#include "datarep.h"
#include "ms_real.h"
#include "ms_ipc.h"
#include "ms_job.h"
#include "ms_real.c"

/*
  Note that the timeout default below is overridden by mslavedc so that 
  manually started servers will persist throughout a typical session. 
*/
#define DEFAULT_SLAVE_TIMEOUT (60)	/* by default time out in 60 seconds */
#define DEFAULT_NICE (10)	/* use nice 10 by default */

char *me;			/* name of program */
int timeout = DEFAULT_SLAVE_TIMEOUT;	/* timeout */
int niceval = DEFAULT_NICE;	/* nice value */
int use_sockets = 1;

/* Log an error message and exit */
error(s)
char *s;
{
#ifdef HAVE_SYSLOG
	syslog(LOG_ERR, "%s: %s (%m)", me, s);
#endif
	exit(1);
}

#ifdef HAVE_SOCKETS

#define RECV(fd, data, len, opts, name, namelen) \
  (use_sockets ? \
     recvfrom(fd, data, len, opts, name, namelen) : \
     read(fd, data, len))

#define SEND(fd, data, len, opts, name, namelen) \
  (use_sockets ? \
     sendto(fd, data, len, opts, name, namelen) : \
     write(fd, data, len))

#else				/* !HAVE_SOCKETS */

#define RECV(fd, data, len, opts, name, namelen) \
  read(fd, data, len)

#define SEND(fd, data, len, opts, name, namelen) \
  write(fd, data, len)

#endif				/* !HAVE_SOCKETS */

/* Decode a single parameter */

real decode_parm(p)
char **p;
{
	real r = net_to_real(ntohli(*((netreal *) * p)));
	*p += sizeof(netreal);
	return r;
}

unsigned int mandelbrot(parms, maxiter, flags)
real *parms;
unsigned int maxiter;
int flags;
{
	register real x_re, x_im;
	register real c_re, c_im;
	register real xresq, ximsq;
	complex c0;
	complex z0;
	real old_re = zero_real(), old_im = zero_real();
	int show_interior;
	unsigned int count;

	show_interior = ! !(flags & MS_OPT_INTERIOR);

	c0.re = parms[0];
	c0.im = parms[1];
	z0.re = parms[2];
	z0.im = parms[3];

	c_re = c0.re;
	c_im = c0.im;
	x_re = z0.re;
	x_im = z0.im;

	/* The following loop is where the Real Work gets done. */
	count = 0;
	while (count < maxiter - 1) {
		/*
		   The following if statement implements limit cycle detection
		   to speed up calculation of areas inside the Mandelbrot set. 
		   Unfortunately, it also slows down the calculation of other areas,
		   to the degree that it probably pays off only when doing shallow
		   zooms with large black areas on machines with slow multiply
		   instructions.  Therefore it is now disabled by default.
		 */

#ifdef CYCLE_DETECT
		if ((count & (count - 1)) == 0) {	/* "count" is zero or a power of two; save the current position */
			old_re = x_re;
			old_im = x_im;
		} else {
			/*
			   Check if we have returned to a previously saved position; 
			   if so, the iteration has converged to a limit cycle => we 
			   are inside the Mandlebrot set and need iterate no further.
			 */
			if (x_re == old_re && x_im == old_im) {
				if (!show_interior)
					count = maxiter - 1;
				break;
			}
		}
#endif
		/* 
		   This is the familiar "z := z^2 + c; abort if |z| > 2"
		   Mandelbrot iteration, with the arithmetic operators hidden
		   in macros so that the same code can be compiled for either
		   fixed-point or floating-point arithmetic.
		   The macros (mul_real(), etc.) are defined in ms_real.h. 
		 */
		xresq = mul_real(x_re, x_re);
		ximsq = mul_real(x_im, x_im);
		if (gteq_real(add_real(xresq, ximsq), four_real()))
			break;
		x_im = add_real(twice_mul_real(x_re, x_im), c_im);
		x_re = add_real(sub_real(xresq, ximsq), c_re);
		count++;
	}
	return count;
}

/* 
  An alternative fractal.  This code is turned of for now because the
  protocol doesn't yet support multiple fractal types.  
*/

#ifdef HENON
unsigned int henon(parms, maxiter, flags)
real *parms;
unsigned int maxiter;
int flags;
{
	register real c0, c1;	/* coefficients */
	register real x0, x1, x2;	/* states */

	unsigned int count;

	c0 = parms[0];
	c1 = parms[1];
	x1 = parms[2];
	x2 = parms[3];

	count = 0;
	while (count < maxiter - 1) {
		x0 = add_real(one_real(),
			      add_real(mul_real(c0, x2),
				       mul_real(c1, mul_real(x1, x1))));
		x2 = x1;
		x1 = x0;

		if (x0 > int_to_real(10) || x0 < int_to_real(-10))
			break;

		count++;
	}
	return count;
}
#endif				/* HENON */

#define MAX_PARMS 32

/*
  Do the actual calculation and encode the results in an reply message;
  return the size of the message (in bytes) 
*/

unsigned int calculate(in, out)
Message *in, *out;
{
	int julia;		/* true if calculating a Julia set */
	int show_interior;	/* true if displaying interior structure */
	complex delta;

	int xc, yc, xmin, xmax, xsize, ymin, ymax, ysize;
	unsigned int maxiter;
	unsigned long mi_count = 0;
	unsigned int datasize;
	int flags;
	ms_job *job;
	char *p;		/* untyped pointer to parameter block in message */
	int x_parm_no, y_parm_no;	/* indices of parameters to vary with x/y coord */
	real *varx, *vary;	/* pointers to same */
	real initial_varx;
	real parm_buf[MAX_PARMS];
	int n_parms;
	int i;

#define parms parm_buf

	job = (ms_job *) in->whip.data;

	/* check that the format is supported */
	if (ntohs(in->whip.header.format) != DATA_FORMAT)
		return (0);

	/* convert values in the message to host byte order and precalculate some */
	/* useful values */
	xmin = ntohs(job->s.x);
	xsize = ntohs(job->s.width);
	xmax = xmin + xsize;
	ymin = ntohs(job->s.y);
	ysize = ntohs(job->s.height);
	ymax = ymin + ysize;

	maxiter = ntohli(job->j.iteration_limit);

	datasize = (maxiter > 256) ?
	    ((char *)(&(out->reply.data.shorts[xsize * ysize]))
	     - (char *)(&(out->reply))) :
	    ((char *)(&(out->reply.data.chars[xsize * ysize]))
	     - (char *)(&(out->reply)));

	/*
	   Perform a simple sanity check to avoid getting into semi-infinite 
	   loops because of malicious or erroneous messages.
	 */
	if (datasize > MAX_DATAGRAM)
		error("data too large");

	/* 
	   Make sure the iteration counts can be represented in the result
	   packet; 65536 iterations can still take a long time and if we
	   keep calculating the same packet for too long the timeout alarm will
	   get us.  Those who need lots of iterations can use smaller chunks.
	 */
	if (maxiter >= 65536)
		error("iteration count too large");

	flags = ntohs(job->j.flags);
	julia = ntohs(job->j.julia);

	if (julia) {
		n_parms = 4;
		x_parm_no = 2;	/* z0.re */
		y_parm_no = 3;	/* z0.im */
	} else {		/* Mandelbrot */

		n_parms = 4;
		x_parm_no = 0;	/* c0.re */
		y_parm_no = 1;	/* c0.im */
	}

	/* point to beginning of parameters */
	p = (char *)&job->j.corner.re;

	/* convert each parameter from network format to computational format */
	for (i = 0; i < n_parms; i++)
		parm_buf[i] = decode_parm(&p);

	/* set up pointers to the parameters that correspond to x and y */
	varx = &parm_buf[x_parm_no];
	vary = &parm_buf[y_parm_no];

	delta.re = decode_parm(&p);
	delta.im = decode_parm(&p);

	/* take the chunk offset into account */
	(*varx) = add_real((*varx), mul_real_int(delta.re, xmin));
	(*vary) = add_real((*vary), mul_real_int(delta.im, ymin));

	/* save initial x coordinate for reuse on subsequent scanlines */
	initial_varx = (*varx);

	for (yc = ymin; yc < ymax; yc++) {
		(*varx) = initial_varx;

		for (xc = xmin; xc < xmax; xc++) {
#ifdef HENON
			unsigned int count = henon(parm_buf, maxiter, flags);
#else
			unsigned int count =
			    mandelbrot(parm_buf, maxiter, flags);
#endif

			if (maxiter > 256)
				out->reply.data.shorts[(yc - ymin) * xsize +
						       (xc - xmin)] =
				    htons(count);
			else
				out->reply.data.chars[(yc - ymin) * xsize +
						      (xc - xmin)] = count;
			mi_count += count;

			(*varx) = add_real((*varx), delta.re);
		}
		(*vary) = add_real((*vary), delta.im);
	}
	out->reply.mi_count = htonl(mi_count);
	return (datasize);
}

void serve()
{
	Message in;
	Message out;
	int isock = 0;
	int osock = use_sockets ? 0 : 1;

	NET_ADDRESS oname;

	while (1) {
		unsigned int bytes;
		int onamelen = sizeof(oname);
		int version;
		if (timeout != 0)
			alarm(timeout);
		/* receive from anywhere, save the address of the caller in oname */
		if (RECV(isock, (char *)&in, sizeof(in), 0,
			 (struct sockaddr *)&oname, &onamelen) < 0)
			error("receiving datagram packet");
		version = ntohs(in.generic.header.version);
		if (ntohs(in.generic.header.magic) == MAGIC
		    && version == VERSION) {
			switch (ntohs(in.generic.header.type)) {
			case WHIP_MESSAGE:
				/* copy the header and id structures as such while still in */
				/* network byte order (not strictly portable but probably works) */
				out.reply.header = in.whip.header;
				out.reply.id = in.whip.id;
				/* just the message type needs to be changed */
				out.reply.header.type = htons(REPLY_MESSAGE);
				/* calculate() sets the data and mi_count fields */
				bytes = calculate(&in, &out);
				if (bytes)
					if (SEND
					    (osock, (char *)&out, (int)bytes, 0,
					     (struct sockaddr *)&oname,
					     sizeof(oname)) < 0)
						error
						    ("sending calculated data");
				break;
			case WHO_R_U_MESSAGE:
				oname.sin_port = in.who.port;	/* in network byte order already */
				out.iam.header = in.who.header;
				out.iam.pid = htons(getpid());
				if (SEND
				    (osock, (char *)&out, sizeof(IAmMessage), 0,
				     (struct sockaddr *)&oname,
				     sizeof(oname)) < 0)
					error
					    ("sending response to pid inquiry");
				break;
			default:;	/* ignore other messages */
			}
		}
	}
}

/*
  This function is called when we get a SIGALRM so that we die gracefully 
  with exit status 0; otherwise inetd would log a message about us getting
  an alarm signal.
*/

int die()
{
	exit(0);
}

int main(argc, argv)
int argc;
char **argv;
{
	static char usage_msg[] = "usage: mslaved [ -options ] mslaved\n";
	me = argv[0];
	signal(SIGALRM, die);
	while (--argc) {
		char *s = *++argv;
		if (*s++ != '-')
			goto usage;
		switch (*s++) {
		case 'n':	/* nice */
			niceval = atoi(*s ? s : (--argc, *++argv));
			break;
		case 't':	/* timeout */
			timeout = atoi(*s ? s : (--argc, *++argv));
			break;
		case 'i':	/* accept for backwards compatibility but ignore */
			break;
		case 'p':	/* pipe mode */
			use_sockets = 0;
			break;
		default:
			goto usage;
		}
	}
	nice(niceval);
	serve();		/* never returns */

 usage:
	/*
	   Try to avoid pulling in printf from the library 
	   (but syslog probably does that anyway...) 
	 */
	write(2, usage_msg, sizeof(usage_msg) - 1);
	exit(1);
}
